downstairs/src/extent_inner_raw.rs

```diff
@@ -0,0 +1,2373 @@
+        // the slots.  It's that range that we'll need to read + write, so we
+        // want to pick whichever slot does less work.
+        #[derive(Copy, Clone)]
+        struct Counter {
+            count: usize,
+            min_block: u64,
+            max_block: u64,
+        }
+        let mut a_count = Counter {
+            count: 0,
+            min_block: u64::MAX,
+            max_block: 0,
+        };
+        let mut b_count = a_count;
+        for (i, s) in self.active_context.iter().enumerate() {
+            let count = match s {
+                ContextSlot::A => &mut a_count,
+                ContextSlot::B => &mut b_count,
+            };
+            count.count += 1;
+            count.min_block = count.min_block.min(i as u64);
+            count.max_block = count.max_block.max(i as u64);
+        }
+        if a_count.count == 0 || b_count.count == 0 {
+            return Ok(());
+        }
+
+        let (copy_from, counter) = if a_count.count < b_count.count {
+            (ContextSlot::A, a_count)
+        } else {
+            (ContextSlot::B, b_count)
+        };
+        assert!(counter.count > 0);
+        assert!(counter.max_block >= counter.min_block);
+        let num_slots = counter.max_block + 1 - counter.min_block;
+
+        // Read the source context slots from the file
+        let source_slots = self.layout.read_context_slots_contiguous(
+            &self.file,
+            counter.min_block,
+            num_slots,
+            copy_from,
+        )?;
+        let mut dest_slots = self.layout.read_context_slots_contiguous(
+            &self.file,
+            counter.min_block,
+            num_slots,
+            !copy_from,
+        )?;
+
+        // Selectively overwrite dest with source context slots
+        for (i, block) in (counter.min_block..=counter.max_block).enumerate() {
+            let block = block as usize;
+            if self.active_context[block] == copy_from {
+                dest_slots[i] = source_slots[i];
+
+                // Mark this slot as unsynched, so that we don't overwrite
+                // it later on without a sync
+                self.context_slot_synched_at[block][!copy_from as usize] =
+                    self.sync_index + 1;
+            }
+        }
+        let r = self.layout.write_context_slots_contiguous(
+            &self.file,
+            counter.min_block,
+            dest_slots.iter().map(|v| v.as_ref()),
+            !copy_from,
+        );
+
+        // If this write failed, then try recomputing every context slot
+        // within the unknown range
+        if r.is_err() {
+            for block in counter.min_block..=counter.max_block {
+                self.recompute_slot_from_file(block).unwrap();
+            }
+        } else {
+            for block in counter.min_block..=counter.max_block {
+                self.active_context[block as usize] = !copy_from;
+            }
+        }
```

Comment:
At this point in the function (if there was no error in writing the context slots), the in memory active context array is different than what active contexts were previously serialized to the file. The second of the following asserts will fire during the defragment tests:

```patch
diff --git a/downstairs/src/extent.rs b/downstairs/src/extent.rs
index 3457512..bbe4cd9 100644
--- a/downstairs/src/extent.rs
+++ b/downstairs/src/extent.rs
@@ -77,7 +77,7 @@ pub(crate) trait ExtentInner: Send + Debug {
 }
 
 /// BlockContext, with the addition of block index and on_disk_hash
-#[derive(Copy, Clone)]
+#[derive(Debug, Copy, Clone, PartialEq)]
 pub struct DownstairsBlockContext {
     pub block_context: BlockContext,

diff --git a/downstairs/src/extent_inner_raw.rs b/downstairs/src/extent_inner_raw.rs
index 0e4f8ea..a35b1ca 100644
--- a/downstairs/src/extent_inner_raw.rs
+++ b/downstairs/src/extent_inner_raw.rs
@@ -1131,6 +1140,10 @@ impl RawInner {
                 self.active_context[block as usize] = !copy_from;
             }
         }
+
+        assert!(!self.dirty);
+        assert_eq!(self.active_context, self.layout.get_active_contexts(&self.file)?);
+
         r.map(|_| ())
     }
 ```

Also the dirty bit is not set (as we just did a flush): if there is a crash now, `Extent::open` will see that the dirty bit is not set and deserialize the active contexts that were previously written to the file, and _not_ validate those against the block's data on disk.

However, I _think_ this is ok: we only defragment after a flush, so we know what that that the block data contents are on durable storage and that the active context slot is correct and matches that. Defragmentation only copies the contents of the active context slot to the inactive context slot, so after defragmentation both slot A and B contain the same (valid) context. The context will be correct for the block data no matter which one is picked.

The following proves that:

```patch
diff --git a/downstairs/src/extent.rs b/downstairs/src/extent.rs
index 3457512..bbe4cd9 100644
--- a/downstairs/src/extent.rs
+++ b/downstairs/src/extent.rs
@@ -77,7 +77,7 @@ pub(crate) trait ExtentInner: Send + Debug {
 }
 
 /// BlockContext, with the addition of block index and on_disk_hash
-#[derive(Copy, Clone)]
+#[derive(Debug, Copy, Clone, PartialEq)]
 pub struct DownstairsBlockContext {
     pub block_context: BlockContext,

diff --git a/downstairs/src/extent_inner_raw.rs b/downstairs/src/extent_inner_raw.rs
index 0e4f8ea..585e18d 100644
--- a/downstairs/src/extent_inner_raw.rs
+++ b/downstairs/src/extent_inner_raw.rs
@@ -1065,6 +1065,13 @@ impl RawInner {
             max_block: 0,
         };
         let mut b_count = a_count;
+
+        // PARANOID: make sure these match
+        assert_eq!(
+            self.active_context,
+            self.layout.get_active_contexts(&self.file)?,
+        );
+
         for (i, s) in self.active_context.iter().enumerate() {
             let count = match s {
                 ContextSlot::A => &mut a_count,
@@ -1086,6 +1093,7 @@ impl RawInner {
         assert!(counter.count > 0);
         assert!(counter.max_block >= counter.min_block);
         let num_slots = counter.max_block + 1 - counter.min_block;
+        assert!(num_slots > 0);
 
         // Read the source context slots from the file
         let source_slots = self.layout.read_context_slots_contiguous(
@@ -1113,6 +1121,7 @@ impl RawInner {
                     self.sync_index + 1;
             }
         }
+
         let r = self.layout.write_context_slots_contiguous(
             &self.file,
             counter.min_block,
@@ -1131,6 +1140,94 @@ impl RawInner {
                 self.active_context[block as usize] = !copy_from;
             }
         }
+
+        // Here, self.active_context does not match what is on disk, and dirty
+        // is not set. If there is a crash now, Extent::open will be called, and
+        // active context will be read from the disk and _not_ validated against
+        // the file data. This is ok if we defragment only after a flush has
+        // successfully occurred: due to how defragmentation works, when there
+        // is a mismatch between what is in memory now and what is on disk, the
+        // two context slots on disk are duplicates of each other, and match the
+        // file data on disk (because we just did a successful flush).
+        //
+        // Go through _every_ block, and debug assert that when there is a
+        // mismatch the two context slots on disk are duplicates of each other,
+        // and match the file data that is on disk.
+        assert!(!self.dirty);
+        assert_ne!(self.active_context, self.layout.get_active_contexts(&self.file)?);
+
+        let block_size = self.extent_size.block_size_in_bytes();
+        for (i, (in_memory_slot, on_disk_slot)) in std::iter::zip(self.active_context.iter(), self.layout.get_active_contexts(&self.file)?.iter()).enumerate() {
+            let block = i as u64;
+
+            let (buf, hash) = {
+                let mut buf = vec![0; block_size as usize];
+
+                nix::sys::uio::pread(
+                    self.file.as_raw_fd(),
+                    &mut buf,
+                    (block_size as u64 * block) as i64,
+                )
+                .map_err(|e| {
+                    CrucibleError::IoError(format!(
+                        "extent {}: reading block {} data failed: {e}",
+                        self.extent_number,
+                        block,
+                    ))
+                })?;
+
+                let hash = integrity_hash(&[&buf]);
+
+                (buf, hash)
+            };
+
+            if in_memory_slot != on_disk_slot {
+                // The in memory active slot does not match the on-disk active
+                // slot. Make sure that the on disk contexts are duplicates, and
+                // that they both match the on disk file data.
+                let a_context = self.layout.read_context_slots_contiguous(
+                    &self.file,
+                    block,
+                    1,
+                    ContextSlot::A,
+                )?;
+                let b_context = self.layout.read_context_slots_contiguous(
+                    &self.file,
+                    block,
+                    1,
+                    ContextSlot::B,
+                )?;
+                debug_assert_eq!(a_context, b_context);
+
+                if let Some(a_context) = a_context[0] {
+                    // If the slot is Some, then there is encrypted data there.
+                    debug_assert_eq!(a_context.block, block);
+                    debug_assert_eq!(a_context.on_disk_hash, hash);
+                } else {
+                    // If the slot is None, then the data should be all zeros
+                    debug_assert_eq!(buf, vec![0u8; block_size as usize]);
+                }
+            } else {
+                // The in memory active slot matches the on disk active slot.
+                // Validate the on disk active slot while we're here.
+                let on_disk_context = self.layout.read_context_slots_contiguous(
+                    &self.file,
+                    block,
+                    1,
+                    *on_disk_slot,
+                )?;
+
+                if let Some(on_disk_context) = on_disk_context[0] {
+                    // If the slot is Some, then there is encrypted data there.
+                    debug_assert_eq!(on_disk_context.block, block);
+                    debug_assert_eq!(on_disk_context.on_disk_hash, hash);
+                } else {
+                    // If the slot is None, then the data should be all zeros
+                    debug_assert_eq!(buf, vec![0u8; block_size as usize]);
+                }
+            }
+        }
+
         r.map(|_| ())
     }
 }
```

What do you think?

Reply:
Excellent analysis, I agree with both parts (the cached context slots on disk disagree with active context slots in-memory, despite dirty being false; and this is still safe, because the cached context slots contain valid contexts)

Reply:
I also added another assertion in 5ed99f7c00aee9b0cfb3f9c4bf7d8d4b42e99959 to make sure no one tries to call `defragment` on a dirty extent.

Reply:
Yup, added a comment in 971ed1e97e24e38bbd24e6991abde4467126b5db